src/stubs/sm_input.c

```diff
@@ -1,25 +1,61 @@
 #include "reactive_stubs_support.h"
 
 #include <alloca.h>
+#include <stdlib.h>
 
-#define AD_SIZE 2
-
-void SM_ENTRY(SM_NAME) __sm_handle_input(uint16_t conn_id,
+void SM_ENTRY(SM_NAME) __sm_handle_input(conn_index conn_id,
                                          const void* payload, size_t len)
 {
-    if (conn_id >= SM_NUM_INPUTS)
-        return;
+    // search for the connection.
+    // Unfortunately, this operation is O(n) with n = number of connections
+    int i;
+    Connection *conn = NULL;
+    for (i=0; i<__sm_num_connections; i++) {
+      conn = &__sm_io_connections[i];
+
+      if(conn->conn_id == conn_id)
+        break;
+    }
 
-    const size_t data_len = len - AD_SIZE - SANCUS_TAG_SIZE;
-    const uint8_t* cipher = (uint8_t*)payload + AD_SIZE;
-    const uint8_t* tag = cipher + data_len;
+    if (i == __sm_num_connections)
+      return; // connection not found
 
-    // TODO check for stack overflow!
-    uint8_t* input_buffer = alloca(data_len);
+    if (conn->io_id >= SM_NUM_INPUTS)
+      return;
 
-    if (sancus_unwrap_with_key(__sm_io_keys[conn_id], payload, AD_SIZE,
-                               cipher, data_len, tag, input_buffer))
-    {
-        __sm_input_callbacks[conn_id](input_buffer, data_len);
+    // associated data only contains the nonce, therefore we can use this
+    // this trick to build the array fastly (i.e. by swapping the bytes)
+    const uint16_t nonce_rev = conn->nonce << 8 | conn->nonce >> 8;
+    const size_t data_len = len - SANCUS_TAG_SIZE;
+
+    if(data_len == 0) {
+      // In this case, sancus_unwrap would always fail due to some bug
+      // therefore we just check if the tag is correct.
+      const uint8_t *tag = payload;
+      uint8_t expected_tag[SANCUS_TAG_SIZE];
+      sancus_tag_with_key(conn->key, &nonce_rev, sizeof(nonce_rev), expected_tag);
+      int success = 1, i;
+      for(i=0; i<SANCUS_TAG_SIZE; i++) {
+        if(tag[i] != expected_tag[i]) {
```

Comment:
don't do this early out! you open a huge timing side channel to leak the expected tag byte per byte! :smiling_imp: 

Instead, either use a proper constant-time memcmp implementation (cf #31 ) or use the trick I employed in VulCAN using `union` types (w the assumption the compiler generates constant-time cmp for union types, which I verified to the case):

https://github.com/sancus-tee/vulcan/blob/master/can-auth/vatican.c#L241

Reply:
let me know if it's unclear or you want to have a call or so to discuss further!

Reply:
As mentioned in another comment, I'd rather remove this code if the issue with `sancus_unwrap` can be fixed in the core. This was meant to be a temporary workaround 😄